wrapper/src/main/java/software/amazon/jdbc/MonitorPluginService.java

```diff
@@ -0,0 +1,795 @@
+      if (hostSpecB.getAvailability() == HostAvailability.AVAILABLE) {
+        changes.add(NodeChangeOptions.WENT_UP);
+      } else if (hostSpecB.getAvailability() == HostAvailability.NOT_AVAILABLE) {
+        changes.add(NodeChangeOptions.WENT_DOWN);
+      }
+    }
+
+    if (!changes.isEmpty()) {
+      changes.add(NodeChangeOptions.NODE_CHANGED);
+    }
+
+    return changes;
+  }
+
+  @Override
+  public List<HostSpec> getAllHosts() {
+    return this.allHosts;
+  }
+
+  @Override
+  public List<HostSpec> getHosts() {
+    AllowedAndBlockedHosts hostPermissions = this.storageService.get(
+        AllowedAndBlockedHosts.class, this.initialConnectionHostSpec.getUrl());
+    if (hostPermissions == null) {
+      return this.allHosts;
+    }
+
+    List<HostSpec> hosts = this.allHosts;
+    Set<String> allowedHostIds = hostPermissions.getAllowedHostIds();
+    Set<String> blockedHostIds = hostPermissions.getBlockedHostIds();
+
+    if (!Utils.isNullOrEmpty(allowedHostIds)) {
+      hosts = hosts.stream()
+          .filter((hostSpec -> allowedHostIds.contains(hostSpec.getHostId())))
+          .collect(Collectors.toList());
+    }
+
+    if (!Utils.isNullOrEmpty(blockedHostIds)) {
+      hosts = hosts.stream()
+          .filter((hostSpec -> !blockedHostIds.contains(hostSpec.getHostId())))
+          .collect(Collectors.toList());
+    }
+
+    return hosts;
+  }
+
+  @Override
+  public void setAvailability(final @NonNull Set<String> hostAliases, final @NonNull HostAvailability availability) {
+
+    if (hostAliases.isEmpty()) {
+      return;
+    }
+
+    final List<HostSpec> hostsToChange = this.getAllHosts().stream()
+        .filter((host) -> hostAliases.contains(host.asAlias())
+            || host.getAliases().stream().anyMatch(hostAliases::contains))
+        .distinct()
+        .collect(Collectors.toList());
+
+    if (hostsToChange.isEmpty()) {
+      LOGGER.finest(() -> Messages.get("PluginServiceImpl.hostsChangelistEmpty"));
+      return;
+    }
+
+    final Map<String, EnumSet<NodeChangeOptions>> changes = new HashMap<>();
+    for (final HostSpec host : hostsToChange) {
+      final HostAvailability currentAvailability = host.getAvailability();
+      host.setAvailability(availability);
+      hostAvailabilityExpiringCache.put(host.getUrl(), availability,
+          DEFAULT_HOST_AVAILABILITY_CACHE_EXPIRE_NANO);
+      if (currentAvailability != availability) {
+        final EnumSet<NodeChangeOptions> hostChanges;
+        if (availability == HostAvailability.AVAILABLE) {
+          hostChanges = EnumSet.of(NodeChangeOptions.WENT_UP, NodeChangeOptions.NODE_CHANGED);
+        } else {
+          hostChanges = EnumSet.of(NodeChangeOptions.WENT_DOWN, NodeChangeOptions.NODE_CHANGED);
+        }
+        changes.put(host.getUrl(), hostChanges);
+      }
+    }
+
+    if (!changes.isEmpty()) {
+      this.pluginManager.notifyNodeListChanged(changes);
+    }
+  }
+
+  @Override
+  public boolean isInTransaction() {
```

Comment:
I don't think that connections in monitoring threads will use transactions. We can simplify implementation here, like do nothing.

Reply:
If we don't expect this to be used and don't want monitoring threads to call it I think we should throw an exception or log a warning and return (I guess return false?). For now I have the unexpected methods in this class throw exceptions so that I can see if they are being called in our integration tests, but we can decide whether we want them to log warnings instead.